cmd/drone-server/drone.go

```diff
@@ -236,6 +244,14 @@ func main() {
 		auth.POST("", server.GetLogin)
 	}
 
+	redirects := r.Group("/redirect")
+	{
+		redirects.Use(server.SetDatastore(store))
+		redirects.Use(server.SetRepo())
+
+		redirects.GET("/:owner/:name/commit/:sha", server.RedirectSha)
```

Comment:
As you can see, all possible build states generated when CI detected

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/views/projects/merge_requests/widget/_heading.html.haml#L3-13

and after initialized javascript https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/views/projects/merge_requests/widget/_heading.html.haml#L27-29

javascript create request to gitlab **merge request controller**m to method **ci_status** 

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/controllers/projects/merge_requests_controller.rb#L179

After, called current **ci_service** ( for example drone ) method **merge_request_status** 

https://gitlab.com/bugagazavr/gitlab-ce/blob/72b32c8e8d82438f5981e3a7075b6c841077c394/app/models/project_services/drone_ci_service.rb#L43

And ajax got a status, than javascript make a visible, previously generated container with status